sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3112,29 +3112,30 @@ case class ArrayDistinct(child: Expression)
     (data: Array[AnyRef]) => new GenericArrayData(data.distinct.asInstanceOf[Array[Any]])
   } else {
     (data: Array[AnyRef]) => {
-      var foundNullElement = false
-      var pos = 0
+      val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
+      var alreadyStoredNull = false
+      var found = false
       for (i <- 0 until data.length) {
-        if (data(i) == null) {
-          if (!foundNullElement) {
-            foundNullElement = true
-            pos = pos + 1
+        if (data(i) != null) {
+          found = false
+          var j = 0;
+          while (!found && j < arrayBuffer.size) {
+            val va = arrayBuffer(j)
+            found = (va != null) && ordering.equiv(va, data(i))
```

Comment:
Rather than handle nulls separately, can you just check it here?
```
found = if (va == null) data(i) == null else ordering.equiv(...)
```
It also kind of looks like the ordering already handles nulls?

Reply:
@srowen Thanks. Actually in my understanding, ordering does not seem to handle nulls.  Given that, the proposed condition does not work  for case when va != null and data[i] is null. We get a null pointer exception. We can probably do something like : 
```
 if (data(i) != null && va != null) {
     found = ordering.equiv(va, data(i))
  } else if (data(i) == null && va == null) {
     found = true
  }
```
Given this, i feel the existing code reads better and i think performs better if we have many null and non-null values in the array by keeping the null handling outside the inner loop. Also, i believe, for other collection operation functions we treat nulls separately. But i will change, if you feel otherwise. Please let me know.

Reply:
You're right, on second look the ordering in ArrayType doesn't handle nulls, only nulls in the array. This is fine. You could save a lookup with...

```
if (data(i) == null) {
  found = va == null
} else if (va != null) {
  found = ordering.equiv(va, data(i))
}
```

but I don't feel strongly about it

Reply:
@dilipbiswal explained my intention in the code of `null` part. @srowen is code simple and easy for reading, but it may include # of iterations if we have already seen `null`.